File: typespec/problem/details.go
// Package problem contains RFC 9457 problem details and Vetchium's stable
// problem-type catalog.
package problem

const MediaType = "application/problem+json"

type Details struct {
	Type     string   `json:"type"`
	Title    string   `json:"title"`
	Status   int      `json:"status"`
	Detail   string   `json:"detail,omitempty"`
	Instance string   `json:"instance,omitempty"`
	Fields   []string `json:"fields,omitempty"`
}

// Body is a problem response, either bare Details or a struct embedding it
// with extension members. Embedding Details promotes ProblemDetails, so any
// such struct satisfies Body without writing the method itself.
type Body interface {
	ProblemDetails() Details
}

func (d Details) ProblemDetails() Details { return d }

var InternalServerError = Details{
	Type:   "vetchium-problem-details/internal-server-error",
	Title:  "Internal Server Error",
	Status: 500,
	Detail: "Server has encountered an error",
}

var InvalidJSONError = Details{
	Type:   "vetchium-problem-details/invalid-json",
	Title:  "Invalid JSON",
	Status: 400,
	Detail: "Request body is not in expected JSON schema",
}

var ValidationFailedError = Details{
	Type:   "vetchium-problem-details/validation-failed",
	Title:  "Validation failed",
	Status: 400,
	Detail: "List of json field names in the request which failed " +
		"validation",
	Fields: []string{},
}

File: typespec/problem/details.ts
/** The media type used for RFC 9457 problem details. */
export const MediaType = "application/problem+json";

/** RFC 9457 problem details returned by API failures. */
export interface Details {
  type: string;
  title: string;
  status: number;
  detail?: string;
  instance?: string;
  fields?: string[];
}

export const InternalServerError: Readonly<Details> = {
  type: "vetchium-problem-details/internal-server-error",
  title: "Internal Server Error",
  status: 500,
  detail: "Server has encountered an error",
};

export const InvalidJSONError: Readonly<Details> = {
  type: "vetchium-problem-details/invalid-json",
  title: "Invalid JSON",
  status: 400,
  detail: "Request body is not in expected JSON schema",
};

export const ValidationFailedError: Readonly<Details> = {
  type: "vetchium-problem-details/validation-failed",
  title: "Validation failed",
  status: 400,
  detail: "List of json field names in the request which failed validation",
  fields: [],
};
